contracts/VaultV3.vy

```diff
@@ -354,121 +379,183 @@ def _issueSharesForAmount(amount: uint256, recipient: address) -> uint256:
 @internal
 def _deposit(_sender: address, _recipient: address, _assets: uint256) -> uint256:
     assert _recipient not in [self, ZERO_ADDRESS], "invalid recipient"
-    
-    shares: uint256 = self._issueSharesForAmount(_assets, _recipient)
+    assets: uint256 = _assets
+
+    if assets == MAX_UINT256:
+        assets = ASSET.balanceOf(_sender)
+
+    assert self._totalAssets() + assets <= self.depositLimit, "exceed deposit limit"
+    assert assets > 0, "cannot deposit zero"
+
+    shares: uint256 = self._issueSharesForAmount(assets, _recipient)
 
-    self.erc20_safe_transferFrom(ASSET.address, msg.sender, self, _assets)
-    self.totalIdle += _assets 
+    self.erc20_safe_transferFrom(ASSET.address, msg.sender, self, assets)
+    self.totalIdle += assets
 
-    log Deposit(_sender, _recipient, _assets, shares)
+    log Deposit(_sender, _recipient, assets, shares)
 
     return shares
 
 @internal
 def _redeem(_sender: address, _receiver: address, _owner: address, _shares: uint256, _strategies: DynArray[address, 10] = []) -> uint256:
-    if _sender != _owner: 
-      self._spendAllowance(_owner, _sender, _shares)
+    if _sender != _owner:
+        self._spendAllowance(_owner, _sender, _shares)
 
     shares: uint256 = _shares
     sharesBalance: uint256 = self.balanceOf[_owner]
 
-    if _shares == MAX_UINT256:
+    if shares == MAX_UINT256:
         shares = sharesBalance
 
     # TODO: is this needed? will revert in burn call
     assert sharesBalance >= shares, "insufficient shares to withdraw"
     assert shares > 0, "no shares to withdraw"
 
-    assets: uint256 = self._convertToAssets(_shares)
+    assets: uint256 = self._convertToAssets(shares)
 
-    # TODO: withdraw from strategies
-    
-    # TODO: gas savings: totalIdle should be cached if used above
-    assert self.totalIdle >= assets, "insufficient total idle"
+    if assets > self.totalIdle:
+        # load to memory to save gas
+        currTotalIdle: uint256 = self.totalIdle
+        currTotalDebt: uint256 = self.totalDebt
 
-    self._burnShares(_shares, _owner)
+        # withdraw from strategies if insufficient total idle
+        assetsNeeded: uint256 = assets - currTotalIdle
+        assetsToWithdraw: uint256 = 0
+        for strategy in _strategies:
+            assert self.strategies[strategy].activation != 0, "inactive strategy"
 
+            assetsToWithdraw = min(assetsNeeded, IStrategy(strategy).withdrawable())
+            # continue if nothing to withdraw
+            if assetsToWithdraw == 0:
+                continue
+
+            IStrategy(strategy).freeFunds(assetsToWithdraw)
+            ASSET.transferFrom(strategy, self, assetsToWithdraw)
```

Comment:
add 
_TODO: check that `assetsToWithdraw` have been released by using `min(ASSET.balanceOf(strategy), assetsToWithdraw)`_?